redbot/core/utils/__init__.py

```diff
@@ -255,3 +256,54 @@ def bounded_gather(
     tasks = (_sem_wrapper(semaphore, task) for task in coros_or_futures)
 
     return asyncio.gather(*tasks, return_exceptions=return_exceptions)
+
+
+class AsyncGen(AsyncIterable[_T]):
+    """Yield entry every `delay` seconds per `steps`."""
+
+    def __init__(
+        self, contents: Union[Iterator[_T], Iterable[_T]], delay: float = 0.0, steps: int = 1
+    ) -> None:
+        self.delay = delay
+        self.content = iter(contents)
+        self.i = 0
+        self.steps = steps
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self) -> Awaitable[_T]:
+        try:
+            i = next(self.content)
+        except StopIteration:
+            raise StopAsyncIteration
+        self.i += 1
+        if self.i % self.steps == 0:
+            await asyncio.sleep(self.delay)
+        return i
+
+    def __await__(self):
+        return self.flatten()
+
+    async def flatten(self) -> List[_T]:
+        return [item async for item in self]
+
+    async def filter(
+        self, function: Union[AsyncIterable[_T], Iterable[_T]], *args, **kwargs
+    ) -> AsyncIterator[_T]:
+        function = partial(function, *args, **kwargs)
+        async for i in async_filter(self, function):
+            yield i
```

Comment:
- Arguments are passed in wrong order to `async_filter()`
- `function`'s type hint is incorrect
- I think `*args, **kwargs` are unnecessary here, if someone wants to, they can pass partial themselves.
- This whole function can be simplified to regular function returning result of `async_filter()` function:
```suggestion
    def filter(self, function: Callable[[_T], Union[bool, Awaitable[bool]]]) -> AsyncFilter[_T]:
        return async_filter(function, self)
```